lib/ansible/plugins/inventory/__init__.py

```diff
@@ -418,7 +420,9 @@ def _add_host_to_keyed_groups(self, keys, variables, host, strict=False, fetch_h
                     trailing_separator = keyed.get('trailing_separator')
                     if trailing_separator is not None and default_value_name is not None:
                         raise AnsibleParserError("parameters are mutually exclusive for keyed groups: default_value|trailing_separator")
-                    if key or (key == '' and default_value_name is not None):
+
```

Comment:
simpler flow?

```
if not key and default_value)name is not None:
    key = default_value_name
```

then its just `if key:` and a few less branches downstream

Reply:
The if branch si the one actually setting the groups and using the default value if present. It needs to run when both the key expression results in a value or a default value is specified. It also still needs to run the checks for each list value in case key resulted in a list.

I didn't use a truthy check and just kept it with `is None` or `== ''` for using the default value because it would treat `0` as `False` and use the default value whereas we currently error. I'm not sure if we should just `str()` the result anyway to handle `int` or other type values but that's another discussion.